firedrake/preconditioners/fdm.py

```diff
@@ -0,0 +1,1036 @@
+from functools import lru_cache, partial
+
+from pyop2 import op2
+from pyop2.sparsity import get_preallocation
+
+from ufl import FiniteElement, VectorElement, TensorElement
+from ufl import FacetNormal, Jacobian, JacobianDeterminant, JacobianInverse
+from ufl import as_tensor, diag_vector, dot, dx, indices, inner, inv
+from ufl.algorithms.ad import expand_derivatives
+
+from firedrake.petsc import PETSc
+from firedrake.preconditioners.base import PCBase
+from firedrake.preconditioners.patch import bcdofs
+from firedrake.preconditioners.pmg import get_permuted_map, tensor_product_space_query
+from firedrake.utils import IntType_c
+from firedrake.dmhooks import get_function_space, get_appctx
+import firedrake.dmhooks as dmhooks
+import firedrake
+import numpy
+
+try:
+    from scipy.linalg import eigh
+
+    def sym_eig(A, B):
+        return eigh(A, B)
+except ImportError:
```

Comment:
Is there much of a performance penalty to using numpy instead of scipy? Should the code warn here that scipy is not available?

Reply:
This is a one time setup, so this should not affect runtime too much. 
I could not find a nice numpy way to apply the inverse of a triangular matrix other than the explicit inverse approach,
but doing so should not exceed the O(N^3) of calling eigh.